### PR TITLE
feat(addon): #291 Setup ping for tab tracking telemetry.

### LINKED ISSUE #291: Perform XHR ping with telemetry

### PR DESCRIPTION
**Pull request by emtwo**
Fixes #291

### CODE REVIEW

File: lib/TelemetrySender.js [hunk 1]
```diff
@@ -1,19 +1,48 @@
 /* globals Services */
 
-const {Cu} = require("chrome");
+const {Cu, Cc, Ci} = require("chrome");
+
 Cu.import("resource://gre/modules/Services.jsm");
+Cu.importGlobalProperties(["XMLHttpRequest"]);
```
Comment (oyiptong) [inline lib/TelemetrySender.js]: how about we use `fetch` instead?

Comment (oyiptong) [inline lib/TelemetrySender.js]: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch

File: lib/TelemetrySender.js [hunk 2]
```diff
@@ -1,19 +1,48 @@
...
+
+function TelemetrySender(pingPref, pingEndpoint) {
+  this._pingPref = pingPref;
 
-function TelemetrySender() {
   Services.obs.addObserver(this, "tab-session-complete");
+  this._prefs = Cc["@mozilla.org/preferences-service;1"]
```
Comment (oyiptong) [inline lib/TelemetrySender.js]: `Preferences.jsm` will be easier to use: https://dxr.mozilla.org/mozilla-central/source/toolkit/modules/Preferences.jsm

Comment (oyiptong) [inline lib/TelemetrySender.js]: an example in action: https://dxr.mozilla.org/mozilla-central/source/browser/components/nsBrowserGlue.js#408

File: lib/TelemetrySender.js [hunk 3]
```diff
@@ -1,19 +1,48 @@
...
+  observe: function(subject, topic, data) {
     if (topic === "tab-session-complete") {
-      // TODO: send ping to server.
+      this._sendPing(data);
     }
   },
 
+  _sendPing: function(data) {
+    // Create a new XMLHttpRequest that is anonymous, i.e., doesn't send cookies
+    let ping = new XMLHttpRequest({mozAnon: true});
```
Comment (oyiptong) [inline lib/TelemetrySender.js]: fetch will make this code much simpler

File: lib/main.js [hunk 4]
```diff
@@ -8,6 +8,9 @@ const {Cu} = require("chrome");
 Cu.import("resource://gre/modules/ClientID.jsm");
 Cu.import("resource://gre/modules/Task.jsm");
 
+const PING_PREF = "browser.activity-streams.ping";
```
Comment (oyiptong) [inline lib/main.js]: Why have these both as constants and as prefs? We should have one or the other, not both.
I'm in favor of hardcoding this and eschewing the use of prefs entirely.

**Comment by oyiptong** [comment]
Good feature. I'd recommend the use of fetch as it's a nicer API.

We need to choose between prefs and consts, but not both. You can set a pref in `package.json`. It will populate the pref on addon startup.

We need to make sure to remove the pref on addon uninstall (but not on upgrade/downgrade/disable etc).

Should you choose to use a hardcoded const, lib/main.js is a good location and might be the simples for now.

File: lib/main.js [hunk 5]
```diff
@@ -8,6 +8,8 @@ const {Cu} = require("chrome");
 Cu.import("resource://gre/modules/ClientID.jsm");
 Cu.import("resource://gre/modules/Task.jsm");
 
+const PING_ENDPOINT = "https://onyx_tiles.stage.mozaws.net/v3/links/activity-stream";
```
Comment (pdehaan) [inline lib/main.js]: Should this be in a env-specific config or something? It seems weird that we'd always send data to a staging server from a prod environment.

Comment (oyiptong) [inline lib/main.js]: I agree, we should set it in `package.json` as a default pref.

e.g.

```javascript
  "preferences": [
    {
      "name": "telemetry.ping.endpoint",
      "title": "Telemetry server endpoint",
      "type": "string",
      "value": "https://onyx_tiles.stage.mozaws.net/v3/links/activity-stream"
    }
  ]
```

Here's one can use `simplePrefs`:

```javascript
const simplePrefs = require("sdk/simple-prefs");

simplePrefs.prefs["telemetry.ping.endpoint"];
```

For development, we probably want to give that pref another value. We should use the `dev-prefs.json` config file.
JPM commands for test and for command line use (`npm run once` etc.) should use `dev-prefs.json`.

e.g.
`$ jpm run --prefs ./dev-prefs.json`

We should amend existing npm scripts to make that work. The contents of `dev-prefs.json` could be the staging server.

@ncloudioj what would be the two endpoints we need here for staging and prod?

Comment (oyiptong) [inline lib/main.js]: One more note about the endpoints, the test environment should not send any telemetry pings.

**Comment by oyiptong** [comment]
Good patch. we're almost there. It just needs some more work with the prefs/runtime parameters and then it's a go!

File: lib/main.js (hunk 5 above)
Comment (ncloudioj) [inline lib/main.js]: - stage: https://onyx_tiles.stage.mozaws.net/v3/links/activity-stream
- prod: https://tiles.services.mozilla.com/v3/links/activity-stream

File: package.json [hunk 6]
```diff
@@ -101,11 +101,11 @@
     "start:css": "npm run bundle:css && npm run bundle:css -- -w",
     "start:html": "npm run bundle:html -- --baseUrl http://localhost:1963/",
     "start:server": "live-server data/content --port=1963 --no-browser",
-    "firefox": "jpm run -b beta",
+    "firefox": "jpm run -b beta --prefs ./dev-prefs.json",
```
Comment (oyiptong) [inline package.json]: you might need to change yamscripts.yml and run `npm run yamscripts` to regenerate the package.json for the tasks

Comment (oyiptong) [inline package.json]: you did that. nevermind

**Comment by oyiptong** [comment]
r+

**Pull request merged**